### PR TITLE
[feat] Expand TornadoVM profiler with GPU power metrics from oneAPI LevelZero SYSMAN API

### PR DESCRIPTION
**Pull request by stratika**
#### Description

This PR implements a new feature regarding adding the power consumption as a metric in the TornadoVM profiler when TornadoVM operates with the `SPIR-V` backend and the `LevelZero` runtime. More specifically, this PR leverages the implementation of power functions through the LevelZero SYSMAN API in the levelzero-jni repository (See [PR8](https://github.com/beehive-lab/levelzero-jni/pull/8)), and implements the [PowerMetric](https://github.com/beehive-lab/TornadoVM/blob/master/tornado-drivers/drivers-common/src/main/java/uk/ac/manchester/tornado/drivers/common/power/PowerMetric.java) interface for the SPIR-V backend. 

The power metric is calculated based on the reading of the energy counters in two timestamps via the methods `readInitialCounters()` and `readFinalCounters()`: i) before the execution of a kernel, ii) right after the execution of a kernel. If the compute function is not heavy the energy counters may be the same, which will result in a power consumption near to zero. Therefore the TornadoVM profiler will report "n/a".

In the case that the dispatching of the code occurs via the OpenCL runtime, the power metrics are not supported and "n/a" is reported by the TornadoVM profiler. The same stands also for level-zero integrated GPU devices. See relevant [issue](https://github.com/intel/compute-runtime/issues/751) reported in the level-zero repository.

**Note:** This is supported on Discrete GPUs (e.g., Intel ARC A770) and users should explicitly export the variable `ZES_ENABLE_SYSNAM` to enable the utilization of the low-level system management API, as per the level-zero specification.
```bash
export ZES_ENABLE_SYSMAN=1
```

#### Problem description

It is a feature related to the TornadoVM profiler, regarding reporting the power consumed during the execution of a compute method (i.e., kernel) on an accelerator. 
 
#### Backend/s tested

Mark the backends affected by this PR.

- [ ] OpenCL
- [ ] PTX
- [x] SPIRV

#### OS tested

Mark the OS where this PR is tested.

- [x] Linux
- [ ] OSx
- [x] Windows

#### Did you check on FPGAs?

If it is applicable, check your changes on FPGAs.

- [ ] Yes
- [x] No

#### How to test the new patch?

Remove the `level-zero` and `levelzero-jni` repositories, to ensure that you use the latest versions.
Current PR is tested with the following commit points:
`level-zero`: [24ad9d6](https://github.com/oneapi-src/level-zero/commit/24ad9d613a6df5613683f5b70a440cd0a732de7b)
`levelzero-jni`: [614855c](https://github.com/beehive-lab/levelzero-jni/commit/614855c54913d94ccd822b6c0b1ebed570746e0d)

```bash
rm -rf level-zero
rm -rf levelzero-jni
make BACKEND=spirv
tornado --enableProfiler console -m tornado.examples/uk.ac.manchester.tornado.examples.VectorAddInt --params="100000"
```

and the profiler unit-tests:

```bash
tornado  --jvm "-Xmx6g -Dtornado.recover.bailout=False -Dtornado.unittests.verbose=True -Dtornado.spirv.dispatcher=levelzero"  -m  tornado.unittests/uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  --params "uk.ac.manchester.tornado.unittests.profiler.TestProfiler"
WARNING: Using incubator modules: jdk.incubator.vector
```

The output is the following:
```bash
{
    "XXX__GENERATED_REDUCE0": {
        "TOTAL_BYTE_CODE_GENERATION": "4342086",
        "TOTAL_CODE_GENERATION_TIME": "9850387",
        "TOTAL_DISPATCH_DATA_TRANSFERS_TIME": "0",
        "TOTAL_TASK_GRAPH_TIME": "94426549",
        "TOTAL_KERNEL_TIME": "7696",
        "TOTAL_GRAAL_COMPILE_TIME": "76017802",
        "TOTAL_DRIVER_COMPILE_TIME": "3609021",
        "COPY_IN_TIME": "20488",
        "COPY_OUT_TIME": "7852",
        "TOTAL_COPY_IN_SIZE_BYTES": "8328",
        "TOTAL_COPY_OUT_SIZE_BYTES": "44",
        "ALLOCATION_BYTES": "4164",
        "XXX__GENERATED_REDUCE0.reduce": {
            "BACKEND": "SPIRV",
            "METHOD": "TestProfiler.reduction",
            "DEVICE_ID": "0:0",
            "DEVICE": "Intel(R) Arc(TM) A770 Graphics",
            "POWER_USAGE_mW": "56979",              # <--- This is the new power metric.
            "TASK_KERNEL_TIME": "4992",
            "TASK_COMPILE_GRAAL_TIME": "68471402",
            "TASK_CODE_GENERATION_TIME": "9266181",
            "TASK_COMPILE_DRIVER_TIME": "3362986"
        },
        "XXX__GENERATED_REDUCE0.reduce_seq0": {
            "BACKEND": "SPIRV",
            "METHOD": "ReduceFactory.rAdd",
            "DEVICE_ID": "0:0",
            "DEVICE": "Intel(R) Arc(TM) A770 Graphics",
            "POWER_USAGE_mW": "n/a",                          # <--- This is the new power metric. The kernel is small according to the kernel time, resulting in obtaining the same energy counters before and after the execution.
            "TASK_KERNEL_TIME": "2704",
            "TASK_COMPILE_GRAAL_TIME": "7546400",
            "TASK_CODE_GENERATION_TIME": "584206",
            "TASK_COMPILE_DRIVER_TIME": "246035"
        }
    }
}
{
    "s0": {
        "TOTAL_BYTE_CODE_GENERATION": "365813",
        "TOTAL_CODE_GENERATION_TIME": "975442",
        "TOTAL_DISPATCH_DATA_TRANSFERS_TIME": "0",
        "TOTAL_TASK_GRAPH_TIME": "11988559",
        "TOTAL_KERNEL_TIME": "3120",
        "TOTAL_GRAAL_COMPILE_TIME": "6897724",
        "TOTAL_DRIVER_COMPILE_TIME": "312736",
        "COPY_IN_TIME": "13104",
        "COPY_OUT_TIME": "6760",
        "TOTAL_COPY_IN_SIZE_BYTES": "176",
        "TOTAL_COPY_OUT_SIZE_BYTES": "88",
        "ALLOCATION_BYTES": "264",
        "s0.t0": {
            "BACKEND": "SPIRV",
            "METHOD": "TestHello.add",
            "DEVICE_ID": "0:0",
            "DEVICE": "Intel(R) Arc(TM) A770 Graphics",
            "POWER_USAGE_mW": "89852",              # <--- This is the new power metric.
            "TASK_KERNEL_TIME": "3120",
            "TASK_COMPILE_GRAAL_TIME": "6897724",
            "TASK_CODE_GENERATION_TIME": "975442",
            "TASK_COMPILE_DRIVER_TIME": "312736"
        }
    }
}
```

----------------------------------------------------------------------------

### CODE REVIEW

File: tornado-drivers/spirv/src/main/java/uk/ac/manchester/tornado/drivers/spirv/SPIRVDeviceContext.java [hunk 1]
```diff
@@ -80,6 +84,11 @@ protected SPIRVDeviceContext(SPIRVDevice device, SPIRVContext context) {
         init(device);
         this.spirvContext = context;
         this.executionIds = Collections.synchronizedSet(new HashSet<>());
+        if (isDeviceContextLevelZero()) {
+            this.powerMetric = new SPIRVLevelZeroPowerMetric(this);
```
Comment (jjfumero) [inline tornado-drivers/spirv/src/main/java/uk/ac/manchester/tornado/drivers/spirv/SPIRVDeviceContext.java:88]: I suggest renaming this field with: 

```suggestion
            this.powerMetricHandler = new SPIRVLevelZeroPowerMetric(this);
```

File: tornado-drivers/spirv/src/main/java/uk/ac/manchester/tornado/drivers/spirv/SPIRVDeviceContext.java [hunk 2]
```diff
@@ -100,6 +109,10 @@ public SPIRVDevice getDevice() {
         return device;
     }
 
+    public PowerMetric getPowerMetric() {
```
Comment (jjfumero) [inline tornado-drivers/spirv/src/main/java/uk/ac/manchester/tornado/drivers/spirv/SPIRVDeviceContext.java:112]: Also the class: 
```suggestion
    public PowerMetricHandler getPowerMetric() {
```

File: tornado-drivers/spirv/src/main/java/uk/ac/manchester/tornado/drivers/spirv/SPIRVDeviceContext.java [hunk 3]
```diff
@@ -149,6 +162,14 @@ public int getDriverIndex() {
         return TornadoRuntimeProvider.getTornadoRuntime().getBackendIndex(SPIRVBackendImpl.class);
     }
 
+    private boolean isDeviceContextLevelZero() {
```
Comment (jjfumero) [inline tornado-drivers/spirv/src/main/java/uk/ac/manchester/tornado/drivers/spirv/SPIRVDeviceContext.java:165]: What if this is done in the concrete instance:

a method within the `SPIRVLevelZeroPowerMetric` and `SPIRVOCLPowerMetric` classes that return true or false accordingly.

**Review by jjfumero — CHANGES_REQUESTED** [review]

**Comment by jjfumero** [comment]
Question: Can TornadoVM perform the export itself?  

An alternative could be to extend the execution plan: `withLevelZeroSysMan() `

File: tornado-drivers/spirv/src/main/java/uk/ac/manchester/tornado/drivers/spirv/SPIRVDeviceContext.java (hunk 3 above)
Comment (stratika) [reply tornado-drivers/spirv/src/main/java/uk/ac/manchester/tornado/drivers/spirv/SPIRVDeviceContext.java:165]: As discussed internally, we will pause this change and add it as part of a refactoring process that will take place later.

**Comment by stratika** [comment]
> Question: Can TornadoVM perform the export itself?
> 
> An alternative could be to extend the execution plan: `withLevelZeroSysMan() `

This is not so easy to implement. In my opinion the `export` should be performed by the user. I extended the profiler documentation page (for readthedocs) to provide user documentation regarding the power usage field.

**Comment by stratika** [comment]
I updated the code based on the comments that I received, and extended the documentation of the profiler. To check the new text you can run:
```bash
make docs
```

The generated HTML pages are in `docs/build/html`.

Let me know of any further thoughts.

**Review by jjfumero — APPROVED** [review]
LGTM

**Review by mairooni — APPROVED** [review]

**Comment by jjfumero** [comment]
Ready to merge

**Pull request merged**